### PR TITLE
Implemented Network Request Retry-Policy in MSAL-Node & Managed Identity

### PR DESCRIPTION
**Pull request by Robbie-Microsoft**
I am waiting to hear feedback on this implementation before I write the unit tests.

Closing in favor of other implementation: https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/6891

### CODE REVIEW

File: lib/msal-node/src/utils/RetryTracker.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ */
+
+import { Logger } from "@azure/msal-common";
+
+export class RetryTracker {
+    private logger: Logger;
+    private retryLimit: number;
+    private retryWaitTimeMilli: number;
+
+    private _numRetried: number;
```
Comment (bgavrilMS) [inline lib/msal-node/src/utils/RetryTracker.ts:13]: inconsistent naming - all private fields should be named in the same way. 
Also, they can probably be readonly?

File: lib/msal-node/src/utils/RetryTracker.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        if (!retryAfterHeader) {
+            this.logger.warning(
+                "The request can not be retried because the original network response contains the Retry-After header."
+            );
+            return false;
+        }
+
+        let retriableStatusCode: boolean = false;
+        switch (statusCode) {
+            case 404: // Not Found
```
Comment (bgavrilMS) [inline lib/msal-node/src/utils/RetryTracker.ts:50]: These should not be hardcoded. Use the ctor.

File: lib/msal-node/src/utils/RetryTracker.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        retryAfterHeader: string | undefined
+    ): Promise<boolean> {
+        if (this._numRetried === this.retryLimit) {
+            this.logger.warning(
+                "The request can not be retried because it has already been retried the maximum number of times."
+            );
+            return false;
+        }
+
+        if (!retryAfterHeader) {
```
Comment (bgavrilMS) [inline lib/msal-node/src/utils/RetryTracker.ts:41]: @neha-bhargava @Avery-Dunn  - what are we doing in MSALs when IMDS returns a RetryAfter header? IMO we should respect this header, but still retry.

File: lib/msal-node/src/utils/RetryTracker.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+import { Logger } from "@azure/msal-common";
+
+export class RetryTracker {
+    private logger: Logger;
+    private retryLimit: number;
+    private retryWaitTimeMilli: number;
+
+    private _numRetried: number;
+    public resetNumRetried(): void {
```
Comment (bgavrilMS) [inline lib/msal-node/src/utils/RetryTracker.ts:14]: Is this needed?

File: lib/msal-node/src/client/ManagedIdentityApplication.ts [hunk 5]
```diff
@@ -46,6 +50,7 @@ export class ManagedIdentityApplication {
     private static nodeStorage: NodeStorage | undefined;
     private networkClient: INetworkModule;
     private cryptoProvider: CryptoProvider;
+    private retryTracker: RetryTracker;
```
Comment (bgavrilMS) [inline lib/msal-node/src/client/ManagedIdentityApplication.ts:53]: Does it need to be a member of this class? It looks to me that it's more of a method variable?

File: lib/msal-node/src/client/ManagedIdentityClient.ts [hunk 6]
```diff
@@ -32,6 +33,7 @@ export class ManagedIdentityClient {
     private nodeStorage: NodeStorage;
     private networkClient: INetworkModule;
     private cryptoProvider: CryptoProvider;
+    private retryTracker: RetryTracker;
```
Comment (bgavrilMS) [inline lib/msal-node/src/client/ManagedIdentityClient.ts:36]: I don't think the lifetime of this object is correct. There should be 1 retry possible for each HTTP call. I think here you will have just 1 try per ManagedIdentityClient, not per request.

File: lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts [hunk 7]
```diff
@@ -154,6 +142,63 @@ export abstract class BaseManagedIdentitySource {
             }
         }
 
+        let serverTokenResponse: ServerAuthorizationTokenResponse =
```
Comment (bgavrilMS) [inline lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts:145]: It seems to me that the RetryTracker object can be created directly in this method only. There's no need to keep it as class member in all these classes.

**Review by bgavrilMS — COMMENTED** [review]
This is a good approach, the "retry tracker" lifetime needs to be adjusted etc. 

I want to propose an alternative, and then we can contrast and compare, knowing that: 

- the retry logic is likely to change (e.g. obey the RetryAfter header)
- it's likely that we will add RetryLogic for other flows, e.g. ConfidentialClient flows, but the logic will be different
- it's likely to we will need to provide app developers a chance to disable retries completely (for them to add their own)



# HttpClientWithRetries

Idea: the retries happen at HTTP layer, so it makes sense for the HTTP layer to handle the retries. 

Http extensibility point (INetworkModule) should not change.

### Scenarios

- MSI retry N times after X amount of time
- MSI retry M times after reading RetryAfter header (not shown here)


### Design

Let's see if we can use the Decorator pattern https://www.tutorialspoint.com/design_pattern/decorator_pattern.htm

See https://gist.github.com/bgavrilMS/ea89ad64c0204193cd62a1a2fca7d50a for the code.

**Comment by bgavrilMS** [comment]
@trwalke - can you also take a look at this design and at my alternative proposal, as we'll want to do smth similar in MSAL.NET as well.

File: lib/msal-node/src/utils/RetryTracker.ts (hunk 3 above)
Comment (neha-bhargava) [reply lib/msal-node/src/utils/RetryTracker.ts:41]: @bgavrilMS I don't think IMDS or any other managed identity endpoints send back a retry-after header.

File: lib/msal-node/src/utils/RetryTracker.ts (hunk 4 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/utils/RetryTracker.ts:14]: If you're talking about the void method, I use it in BaseManagedIdentitySource on line 173 when the retry limit has been reached (if applicable). However I like your idea about making the RetryTracker local to the network request method in BaseManagedIdentitySource which might eliminate the need for this.

File: lib/msal-node/src/client/ManagedIdentityApplication.ts (hunk 5 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/client/ManagedIdentityApplication.ts:53]: You're right, it doesn't need to be.

File: lib/msal-node/src/client/ManagedIdentityClient.ts (hunk 6 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/client/ManagedIdentityClient.ts:36]: Correct, but this is where the resetNumRetried() method comes into play. I reset the counter of retriedRequests to 0 after each request (retried request, if applicable).

File: lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts (hunk 7 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts:145]: I like this idea. My original idea was to create it just once, and use a setter for resetting the number of times retried to 0 after each request (retried request, if applicable).

**Comment by Robbie-Microsoft** [comment]
Closing in favor of other implementation: [HttpClient-Decorator](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/6891)

**Pull request closed without merging**